### ISSUE TITLE
Detached signing: Combining multiple SignedInfo elements to attach to a SOAP header

### ISSUE THREAD

**Issue by Heilage** [XML-Security/signxml#63]
I'm currently trying to use this library to sign an RST request using SOAP1.2/WS-SecureConversation (a subset of WS-SE and WS-Trust). Since there doesn't appear to exist any implementations of this as of today, I have to make it on my own. 

What the RST looks like before signing (redacted some info, but nothing crucial):

``` XML
<soap:Envelope xmlns:soap="http://www.w3.org/2003/05/soap-envelope">
  <soap:Header>
    <Action xmlns="http://www.w3.org/2005/08/addressing">http://schemas.xmlsoap.org/ws/2005/02/trust/RST/SCT</Action>
    <MessageID xmlns="http://www.w3.org/2005/08/addressing">urn:uuid:cc1549c2-3fa8-439a-b12e-8170288a34f7</MessageID>
    <To wsu:Id="9ef68234-476e-4797-8f31-73c0cae42eab" xmlns="http://www.w3.org/2005/08/addressing" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">service_url</To>
    <ReplyTo xmlns="http://www.w3.org/2005/08/addressing">
      <Address>http://www.w3.org/2005/08/addressing/anonymous</Address>
    </ReplyTo>
    <wsse:Security soap:mustUnderstand="true" xmlns:wsse="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">
      <wsse:BinarySecurityToken EncodingType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-soap-message-security-1.0#Base64Binary" ValueType="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-x509-token-profile-1.0#X509v3" wsu:Id="X509-26ac00cb-9651-4e45-ae8f-6d60fc101bcb">#####</wsse:BinarySecurityToken>
      <wsu:Timestamp wsu:Id="TS-f9269060-f3cd-44b9-b55d-08261909ed56">
        <wsu:Created>2016-08-30T09:10:09.177863+00:00</wsu:Created>
        <wsu:Expires>2016-08-30T09:10:09.177863+00:00</wsu:Expires>
      </wsu:Timestamp>
    </wsse:Security>
  </soap:Header>
  <soap:Body>
    <wst:RequestSecurityToken xmlns:wst="http://schemas.xmlsoap.org/ws/2005/02/trust">
      <wst:RequestType>http://schemas.xmlsoap.org/ws/2005/02/trust/Issue</wst:RequestType>
      <wsp:AppliesTo xmlns:wsp="http://www.w3.org/ns/ws-policy">
        <wsa:EndpointReference xmlns:wsa="http://www.w3.org/2005/08/addressing">
          <wsa:Address>service_url</wsa:Address>
        </wsa:EndpointReference>
      </wsp:AppliesTo>
      <wst:Lifetime xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">
        <wst:Created>2016-08-30T09:10:09.178001+00:00</wst:Created>
        <wst:Expires>2016-08-30T09:15:09.178001+00:00</wst:Expires>
      </wst:Lifetime>
      <wst:TokenType>http://schemas.xmlsoap.org/ws/2005/02/sc/sct</wst:TokenType>
      <wst:KeySize>256</wst:KeySize>
      <wst:Entropy>
        <wst:BinarySecret Type="http://schemas.xmlsoap.org/ws/2005/02/trust/Nonce">###</wst:BinarySecret>
      </wst:Entropy>
      <wst:ComputedKeyAlgorithm>http://schemas.xmlsoap.org/ws/2005/02/trust/CK/PSHA1</wst:ComputedKeyAlgorithm>
      <wst:Renewing/>
    </wst:RequestSecurityToken>
  </soap:Body>
</soap:Envelope>
```

Before this request can be dispatched, it needs to be signed. Specifically, the elements `<To>` and `<Timestamp>`. By using the `reference_uri` argument I can pass each element separately, but the problem is that this also produces two ´<SignedInfo>` fields, and the field `<SignatureValue>` will of course not be correctly computed (I also need to customize `<KeyInfo>` after signing, it has to contain a token reference with a SHA1 thumbpring instead of the x509-data, I'm not sure how this affects the validity of the signature?). 

Anyway, any suggestions as to how I can make this work? This is the final piece of the (pretty big) puzzle to send this request, and I really like the structure of this library. I'm not expecting this to be supported as is, but perhaps there could be a workaround?

**Comment by kislyuk** [comment]
Thanks for reporting. Although typically only one "envelope" node is signed with all the data contained under it, the XML Signature standard does provide a method to sign multiple nodes together, but this library does not yet support it. The way to do this would be to create a signature node and put multiple `<Reference>`s in it. (You need to verify with your application/standard (WS-Trust?) to make sure that's what your standard says to do.)

I can try to whip up support for this in the next few days, although I can't give any firm ETA.

Where in the document does the signature get attached? (As a child of `Envelope`, or enveloping?)

**Comment by Heilage** [comment]
Thanks for the quick reply. The method I am describing follows the standard for signing RST messages as far as I can gather, and should be valid provided  `<Signature>` is of the correct format. The signature will be inserted as a child member of `<wsse:Security>`. I'm using the detached method, and need to sign 1..n given Reference  URIs to compile into a single `<Signature>`, with n number of `<Reference>` elements and `<SignatureValue>` calculated by the whole set. I hope I managed to make myself understood.

No expectations of ETAs or anything. I could perhaps give it a go, but I am short on time these days. My initial suggestion, without having delved too deply into this library's code, would be to allow `reference_uri` to be an iterable, and compute the given digest values as you go. But I am not sure if this would even work. Thoughts?

If you wish to know more about the standard, the documentation is here: http://docs.oasis-open.org/ws-sx/ws-secureconversation/200512/ws-secureconversation-1.3-os.html

**Comment by Heilage** [comment]
Have you been able to look into this? I thought perhaps I could give it a go next week if you're not able to figure it out or don't have time.

**Comment by kislyuk** [comment]
Yes, I have started to look into supporting ws-security, but I don't think I'll get to it in the next few weeks, so please feel free to take a stab at it.

There are two nicely separable preliminary bits of work here: support for multiple `Reference` nodes, and support for custom `KeyInfo`.

Custom `KeyInfo` is probably the easiest, you just need to add the `key_info` argument to `verify()`, pass an ElementTree node, and insert it instead of generating a new one in https://github.com/XML-Security/signxml/blob/master/signxml/__init__.py#L337.

Multiple reference nodes is trickier. I agree with the notion of making reference_uri an iterable: `sign(..., reference_uri=["#id1", "#id2", ...])`. You will also have to make verify() gather these references and return a list of the resulting nodes in `VerifyResult.signed_data`.

**Comment by kislyuk** [comment]
You might also want to start thinking about how you're going to securely resolve the key from `wsse:SecurityTokenReference`. One way to do that might be to pass a callable `key_resolver` callback similarly to `uri_resolver`, which will be given a `KeyInfo` node and return a ssl.crypto.x509 or cryptography verifier callable.

**Comment by kislyuk** [comment]
Reference material: https://www.oasis-open.org/committees/download.php/16357/Trust_SecureConversation_Interop.2004-10.doc

**Comment by voidshard** [comment]
Hey @kislyuk , 


May I trouble you for an example on signing SOAP requests with hmac-sha1? They're all "TODO" still 😄 

For some background - I'm trying to do something close to the OP of this thread, hoping to add a sig like
```
<Signature xmlns="http://www.w3.org/2000/09/xmldsig#">
				<SignedInfo>
					<CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#" />
					<SignatureMethod Algorithm="http://www.w3.org/2000/09/xmldsig#hmac-sha1" />
					<Reference URI="#_0">
						<Transforms>
							<Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#" />
						</Transforms>
						<DigestMethod Algorithm="http://www.w3.org/2000/09/xmldsig#sha1" />
						<DigestValue>...</DigestValue>
					</Reference>
				</SignedInfo>
				<SignatureValue>... </SignatureValue>
				<KeyInfo>
					<o:SecurityTokenReference>
						<o:Reference ValueType="http://schemas.xmlsoap.org/ws/2005/02/sc/sct" URI="#uuid-..." />
					</o:SecurityTokenReference>
				</KeyInfo>
			</Signature>
```

Where I've been given a security token after sending "RequestSecurityToken" 

From what I can work out reading WS-Security I'll need to sign the Body, Timestamp(s), To, From, FaultTo and a few other SOAP header fields. If I understand your functions correctly, I can assign Ids to these elements like URI="_0" or Id="_1" and pass in 

```
    signer = signxml.XMLSigner(signxml.methods.detached, "hmac-sha1", "sha1")
    result = signer.sign(
        document,
        key=token,
        key_info=security_token_reference_element,
        reference_uri=["_0", "_1"],
    )
```

No joy thus far
```
/lib/python3.6/site-packages/cryptography/hazmat/backends/openssl/hmac.py", line 36, in __init__
    ctx, key, len(key), evp_md, self._backend._ffi.NULL
TypeError: initializer for ctype 'void *' must be a cdata pointer, not str
```

Which means .. I'm probably doing something wrong? Would you have an example of how one could do this?
